src/main/java/htsjdk/variant/variantcontext/VariantContextBuilder.java

```diff
@@ -323,12 +335,27 @@ public VariantContextBuilder rmAttributes(final List<String> keys) {
      * collection, so methods that want to add / remove records require the attributes to be copied to a
      */
     private void makeAttributesModifiable() {
-        if ( ! attributesCanBeModified ) {
+        if (!attributesCanBeModified) {
             this.attributesCanBeModified = true;
-            if (attributes == null) {
-            	this.attributes = new HashMap<String, Object>();
-            } else {
-            	this.attributes = new HashMap<String, Object>(attributes);
+
+            final Map<String, Object> tempAttributes = attributes;
+            this.attributes = new HashMap<>();
+            if (tempAttributes != null) {
+                this.attributes = new HashMap<>(tempAttributes);
+            }
+        }
+    }
+
+    /**
+     * Makes the filters modifiable.
+     */
+    private void makeFiltersModifiable() {
+        if (!filtersCanBeModified) {
```

Comment:
Instead of `filtersCanBeModified` could you instead always make the filters modifiable? E.g. whenever setting filters always create a copy, so you don't have to worry about making it modifiable later on. Or is the desire here to avoid copying `filters` unless you have to, e.g. due to memory or other concerns?

Reply:
yes, this is an optimization, since the normal pattern doesn't involve changing the filters after instantiating...